CUDA link issue with incremental build

#848 seems to break incremental builds. Building celeritas from scratch will compile and link everything without issues. Changing a cuda-related file - e.g. `ActionLauncher.device.hh`, even just adding a blank line - then trying to recompile and link will result in the linker error below. This wasn't an issue in 9e1d185.

@sethrj @drbenmorgan @pcanal 

```
[36/43] Linking CXX executable bin/celer-sim
FAILED: bin/celer-sim 
: && /cvmfs/sft.cern.ch/lcg/releases/gcc/11.3.0-ad0f5/x86_64-centos7/bin/g++ -Wall -Wextra -Wno-psabi -pedantic -pedantic-errors -O3 -DNDEBUG -march=cascadelake -mtune=cascadelake  app/celer-sim/CMakeFiles/celer-sim.dir/celer-sim.cc.o app/celer-sim/CMakeFiles/celer-sim.dir/Runner.cc.o app/celer-sim/CMakeFiles/celer-sim.dir/RunnerOutput.cc.o app/celer-sim/CMakeFiles/celer-sim.dir/Transporter.cc.o app/celer-sim/CMakeFiles/celer-sim.dir/RunnerInputIO.json.cc.o app/celer-sim/CMakeFiles/celer-sim.dir/RootOutput.cc.o -o bin/celer-sim  -Wl,-rpath,/home/esseivaj/bld4/devel/spack/var/spack/environments/celeritas/.spack-env/view/lib/root:/home/esseivaj/bld4/devel/celeritas/build-ndebug/lib64:/home/esseivaj/bld4/devel/spack/var/spack/environments/celeritas/.spack-env/view/lib64::  /home/esseivaj/bld4/devel/spack/var/spack/environments/celeritas/.spack-env/view/lib/root/libTree.so.6.28.04  lib64/libceleritas_final.so  /cvmfs/sft.cern.ch/lcg/releases/gcc/11.3.0-ad0f5/x86_64-centos7/lib64/libgomp.so  -lpthread  /home/esseivaj/bld4/devel/spack/var/spack/environments/celeritas/.spack-env/view/lib/root/libNet.so.6.28.04  /home/esseivaj/bld4/devel/spack/var/spack/environments/celeritas/.spack-env/view/lib/root/libRIO.so.6.28.04  /home/esseivaj/bld4/devel/spack/var/spack/environments/celeritas/.spack-env/view/lib/root/libThread.so.6.28.04  -lpthread  /home/esseivaj/bld4/devel/spack/var/spack/environments/celeritas/.spack-env/view/lib/root/libMathCore.so.6.28.04  /home/esseivaj/bld4/devel/spack/var/spack/environments/celeritas/.spack-env/view/lib/root/libCore.so.6.28.04  lib64/libceleritas.so  /home/esseivaj/bld4/devel/spack/var/spack/environments/celeritas/.spack-env/view/lib64/libvecgeomcuda.so  lib64/libcorecel.so  /home/esseivaj/bld4/devel/spack/var/spack/environments/celeritas/.spack-env/view/lib64/libvecgeom.so  -lrt  -lpthread  -Wl,-rpath-link,/home/esseivaj/bld4/devel/spack/var/spack/environments/celeritas/.spack-env/view/lib64 && :
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.37-355ed/x86_64-centos7/bin/ld: lib64/libceleritas_final.so: undefined reference to `__fatbinwrap_f9193967_17_BoundaryAction_cu_7847f3ed_23182'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.37-355ed/x86_64-centos7/bin/ld: lib64/libceleritas_final.so: undefined reference to `__fatbinwrap_78cc259c_24_MuBremsstrahlungModel_cu_7847f3ed_23116'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.37-355ed/x86_64-centos7/bin/ld: lib64/libceleritas.so: undefined reference to `__cudaRegisterLinkedBinary_d62f876c_31_AlongStepRZMapFieldMscAction_cu_7847f3ed_28614'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.37-355ed/x86_64-centos7/bin/ld: lib64/libceleritas.so: undefined reference to `__cudaRegisterLinkedBinary_74d9648b_19_AlongStepKernels_cu_7847f3ed_28441'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.37-355ed/x86_64-centos7/bin/ld: lib64/libceleritas_final.so: undefined reference to `__fatbinwrap_30326de6_20_CombinedBremModel_cu_7847f3ed_23003'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.37-355ed/x86_64-centos7/bin/ld: lib64/libceleritas.so: undefined reference to `__cudaRegisterLinkedBinary_7d3cad3b_28_ExtendFromPrimariesAction_cu_7847f3ed_28701'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.37-355ed/x86_64-centos7/bin/ld: lib64/libceleritas_final.so: undefined reference to `__fatbinwrap_0b145eb6_16_RayleighModel_cu_7847f3ed_23125'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.37-355ed/x86_64-centos7/bin/ld: lib64/libceleritas.so: undefined reference to `__cudaRegisterLinkedBinary_853ff234_24_RelativisticBremModel_cu_7847f3ed_28562'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.37-355ed/x86_64-centos7/bin/ld: lib64/libceleritas.so: undefined reference to `__cudaRegisterLinkedBinary_c1895311_20_MollerBhabhaModel_cu_7847f3ed_28439'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.37-355ed/x86_64-centos7/bin/ld: lib64/libceleritas.so: undefined reference to `__cudaRegisterLinkedBinary_14eea905_30_ExtendFromSecondariesAction_cu_7847f3ed_28705'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.37-355ed/x86_64-centos7/bin/ld: lib64/libceleritas_final.so: undefined reference to `__fatbinwrap_9ea30712_28_AlongStepUniformMscAction_cu_7847f3ed_23269'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.37-355ed/x86_64-centos7/bin/ld: lib64/libceleritas_final.so: undefined reference to `__fatbinwrap_9e1fdb60_19_ActionDiagnostic_cu_7847f3ed_23457'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.37-355ed/x86_64-centos7/bin/ld: lib64/libceleritas.so: undefined reference to `__cudaRegisterLinkedBinary_b35ed0ed_17_StepDiagnostic_cu_7847f3ed_28727'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.37-355ed/x86_64-centos7/bin/ld: lib64/libceleritas_final.so: undefined reference to `__fatbinwrap_85e8a02e_15_EPlusGGModel_cu_7847f3ed_23009'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.37-355ed/x86_64-centos7/bin/ld: lib64/libceleritas.so: undefined reference to `__cudaRegisterLinkedBinary_30326de6_20_CombinedBremModel_cu_7847f3ed_28443'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.37-355ed/x86_64-centos7/bin/ld: lib64/libceleritas.so: undefined reference to `__cudaRegisterLinkedBinary_78cc259c_24_MuBremsstrahlungModel_cu_7847f3ed_28483'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.37-355ed/x86_64-centos7/bin/ld: lib64/libceleritas.so: undefined reference to `__cudaRegisterLinkedBinary_facc871a_21_SeltzerBergerModel_cu_7847f3ed_28570'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.37-355ed/x86_64-centos7/bin/ld: lib64/libceleritas_final.so: undefined reference to `__fatbinwrap_b42bb3d4_25_InitializeTracksAction_cu_7847f3ed_23442'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.37-355ed/x86_64-centos7/bin/ld: lib64/libceleritas_final.so: undefined reference to `__fatbinwrap_d62f876c_31_AlongStepRZMapFieldMscAction_cu_7847f3ed_23297'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.37-355ed/x86_64-centos7/bin/ld: lib64/libceleritas.so: undefined reference to `__cudaRegisterLinkedBinary_f9193967_17_BoundaryAction_cu_7847f3ed_28587'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.37-355ed/x86_64-centos7/bin/ld: lib64/libceleritas_final.so: undefined reference to `__fatbinwrap_f2107e2b_20_BetheHeitlerModel_cu_7847f3ed_22994'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.37-355ed/x86_64-centos7/bin/ld: lib64/libceleritas_final.so: undefined reference to `__fatbinwrap_b35ed0ed_17_StepDiagnostic_cu_7847f3ed_23499'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.37-355ed/x86_64-centos7/bin/ld: lib64/libceleritas.so: undefined reference to `__cudaRegisterLinkedBinary_4e3221b4_19_StepGatherAction_cu_7847f3ed_28772'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.37-355ed/x86_64-centos7/bin/ld: lib64/libceleritas_final.so: undefined reference to `__fatbinwrap_4e3221b4_19_StepGatherAction_cu_7847f3ed_23533'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.37-355ed/x86_64-centos7/bin/ld: lib64/libceleritas_final.so: undefined reference to `__fatbinwrap_50663ced_25_AlongStepNeutralAction_cu_7847f3ed_23241'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.37-355ed/x86_64-centos7/bin/ld: lib64/libceleritas.so: undefined reference to `__cudaRegisterLinkedBinary_b7556077_23_DiscreteSelectAction_cu_7847f3ed_28617'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.37-355ed/x86_64-centos7/bin/ld: lib64/libceleritas.so: undefined reference to `__cudaRegisterLinkedBinary_0b145eb6_16_RayleighModel_cu_7847f3ed_28555'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.37-355ed/x86_64-centos7/bin/ld: lib64/libceleritas.so: undefined reference to `__cudaRegisterLinkedBinary_9ea30712_28_AlongStepUniformMscAction_cu_7847f3ed_28592'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.37-355ed/x86_64-centos7/bin/ld: lib64/libceleritas_final.so: undefined reference to `__fatbinwrap_c1895311_20_MollerBhabhaModel_cu_7847f3ed_23063'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.37-355ed/x86_64-centos7/bin/ld: lib64/libceleritas_final.so: undefined reference to `__fatbinwrap_6907ffe6_16_PreStepAction_cu_7847f3ed_23318'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.37-355ed/x86_64-centos7/bin/ld: lib64/libceleritas_final.so: undefined reference to `__fatbinwrap_5bdc085a_20_KleinNishinaModel_cu_7847f3ed_23024'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.37-355ed/x86_64-centos7/bin/ld: lib64/libceleritas_final.so: undefined reference to `__fatbinwrap_74d9648b_19_AlongStepKernels_cu_7847f3ed_22925'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.37-355ed/x86_64-centos7/bin/ld: lib64/libceleritas_final.so: undefined reference to `__fatbinwrap_853ff234_24_RelativisticBremModel_cu_7847f3ed_23144'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.37-355ed/x86_64-centos7/bin/ld: lib64/libceleritas_final.so: undefined reference to `__fatbinwrap_7d3cad3b_28_ExtendFromPrimariesAction_cu_7847f3ed_23409'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.37-355ed/x86_64-centos7/bin/ld: lib64/libceleritas_final.so: undefined reference to `__fatbinwrap_14eea905_30_ExtendFromSecondariesAction_cu_7847f3ed_23437'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.37-355ed/x86_64-centos7/bin/ld: lib64/libceleritas.so: undefined reference to `__cudaRegisterLinkedBinary_5bdc085a_20_KleinNishinaModel_cu_7847f3ed_28440'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.37-355ed/x86_64-centos7/bin/ld: lib64/libceleritas.so: undefined reference to `__cudaRegisterLinkedBinary_6907ffe6_16_PreStepAction_cu_7847f3ed_28702'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.37-355ed/x86_64-centos7/bin/ld: lib64/libceleritas_final.so: undefined reference to `__fatbinwrap_b7556077_23_DiscreteSelectAction_cu_7847f3ed_23298'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.37-355ed/x86_64-centos7/bin/ld: lib64/libceleritas.so: undefined reference to `__cudaRegisterLinkedBinary_9e1fdb60_19_ActionDiagnostic_cu_7847f3ed_28709'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.37-355ed/x86_64-centos7/bin/ld: lib64/libceleritas_final.so: undefined reference to `__fatbinwrap_8aacd802_19_LivermorePEModel_cu_7847f3ed_23059'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.37-355ed/x86_64-centos7/bin/ld: lib64/libceleritas.so: undefined reference to `__cudaRegisterLinkedBinary_85e8a02e_15_EPlusGGModel_cu_7847f3ed_28437'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.37-355ed/x86_64-centos7/bin/ld: lib64/libceleritas.so: undefined reference to `__cudaRegisterLinkedBinary_f2107e2b_20_BetheHeitlerModel_cu_7847f3ed_28438'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.37-355ed/x86_64-centos7/bin/ld: lib64/libceleritas.so: undefined reference to `__cudaRegisterLinkedBinary_8aacd802_19_LivermorePEModel_cu_7847f3ed_28442'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.37-355ed/x86_64-centos7/bin/ld: lib64/libceleritas.so: undefined reference to `__cudaRegisterLinkedBinary_b42bb3d4_25_InitializeTracksAction_cu_7847f3ed_28710'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.37-355ed/x86_64-centos7/bin/ld: lib64/libceleritas_final.so: undefined reference to `__fatbinwrap_facc871a_21_SeltzerBergerModel_cu_7847f3ed_23181'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.37-355ed/x86_64-centos7/bin/ld: lib64/libceleritas.so: undefined reference to `__cudaRegisterLinkedBinary_50663ced_25_AlongStepNeutralAction_cu_7847f3ed_28588'
```

```
: && /srv/software/gnu10/10.2.0/bin/g++ -Wall -Wextra -pedantic -g  test/CMakeFiles/testdetail_TestMacros.dir/TestMacros.test.cc.o -o test/testdetail_TestMacros  -Wl,-rpath,/nashome/p/pcanal/geant/gcc10/builds/scalar-cuda12/celeritas/lib64:/nashome/p/pcanal/geant/gcc10/install/opt/lib64  lib64/libtestcel_harness.so  lib64/libcorecel_final.so  /nashome/p/pcanal/geant/gcc10/install/opt/lib64/libgtest.so.1.10.0  -pthread  lib64/libcorecel.so && :
lib64/libcorecel.so: undefined reference to `__cudaRegisterLinkedBinary_4e0d1928_9_Filler_cu_314c4826_29355'
lib64/libcorecel_final.so: undefined reference to `__fatbinwrap_4e0d1928_9_Filler_cu_438f1dcd_109866'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
pcanal@wc:celeritas $ touch ~/geant/sources/celeritas/src/corecel/data/detail/Filler.cu
pcanal@wc:celeritas $ ninja -j 1
[4/365] Linking CXX executable test/testdetail_TestMacros
FAILED: test/testdetail_TestMacros
: && /srv/software/gnu10/10.2.0/bin/g++ -Wall -Wextra -pedantic -g  test/CMakeFiles/testdetail_TestMacros.dir/TestMacros.test.cc.o -o test/testdetail_TestMacros  -Wl,-rpath,/nashome/p/pcanal/geant/gcc10/builds/scalar-cuda12/celeritas/lib64:/nashome/p/pcanal/geant/gcc10/install/opt/lib64  lib64/libtestcel_harness.so  lib64/libcorecel_final.so  /nashome/p/pcanal/geant/gcc10/install/opt/lib64/libgtest.so.1.10.0  -pthread  lib64/libcorecel.so && :
lib64/libcorecel.so: undefined reference to `__cudaRegisterLinkedBinary_4e0d1928_9_Filler_cu_ce41cbe7_29634'
lib64/libcorecel_final.so: undefined reference to `__fatbinwrap_4e0d1928_9_Filler_cu_438f1dcd_109866'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

--- a/cmake/CeleritasLibrary.cmake
+++ b/cmake/CeleritasLibrary.cmake
@@ -759,6 +759,9 @@ function(celeritas_target_link_libraries target)
             PRIVATE
             $<DEVICE_LINK:$<TARGET_FILE:${_libstatic}>>
           )
+          set_property(TARGET ${_target_final} APPEND
+            PROPERTY LINK_DEPENDS $<TARGET_FILE:${_libstatic}>
+          )
 
           # Also pass on the the options and definitions.
           celeritas_transfer_setting(${_libstatic} ${_target_final} COMPILE_OPTIONS)

--- a/cmake/CeleritasLibrary.cmake
+++ b/cmake/CeleritasLibrary.cmake
@@ -359,6 +359,7 @@ function(celeritas_rdc_add_library target)
   add_library(Celeritas::${target}_final ALIAS ${target}_final)
   set_target_properties(${target}_final PROPERTIES
     ${_common_props}
+    LINK_DEPENDS $<TARGET_FILE:${target}${_staticsuf}>
     CELERITAS_CUDA_LIBRARY_TYPE Final
     CUDA_RESOLVE_DEVICE_SYMBOLS ON
     EXPORT_PROPERTIES "CELERITAS_CUDA_LIBRARY_TYPE;CELERITAS_CUDA_FINAL_LIBRARY;CELERITAS_CUDA_MIDDLE_LIBRARY;CELERITAS_CUDA_STATIC_LIBRARY"

